src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+                )
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Iterator of all installed binaries.
+    /// Items are `(pkg_id, bins)` where `bins` is the set of binaries that
+    /// package installed.
+    pub fn all_installed_bins(&self) -> impl Iterator<Item = (&PackageId, &BTreeSet<String>)> {
+        self.v1.v1.iter()
+    }
+
+    /// Set of binaries installed by a particular package.
+    /// Returns None if the package is not installed.
+    pub fn installed_bins(&self, pkg_id: PackageId) -> Option<&BTreeSet<String>> {
+        self.v1.v1.get(&pkg_id)
+    }
+
+    /// Remove a package from the tracker.
+    pub fn remove(&mut self, pkg_id: PackageId, bins: &BTreeSet<String>) {
+        self.v1.remove(pkg_id, bins);
+        if self.unstable_upgrade {
+            self.v2.remove(pkg_id, bins);
+        }
+    }
+}
+
 impl CrateListingV1 {
-    pub fn v1(&self) -> &BTreeMap<PackageId, BTreeSet<String>> {
-        &self.v1
+    fn package_for_bin(&self, bin_name: &str) -> Option<PackageId> {
+        self.v1
+            .iter()
+            .find(|(_, bins)| bins.contains(bin_name))
+            .map(|(pkg_id, _)| *pkg_id)
     }
 
-    pub fn v1_mut(&mut self) -> &mut BTreeMap<PackageId, BTreeSet<String>> {
-        &mut self.v1
+    fn mark_installed(&mut self, pkg: &Package, bins: &BTreeSet<String>) {
+        // Remove bins from any other packages.
+        for other_bins in self.v1.values_mut() {
+            for bin in bins {
+                other_bins.remove(bin);
+            }
+        }
+        // Remove empty metadata lines. If only BTreeMap had `retain`.
+        let to_remove = self
+            .v1
+            .iter()
+            .filter_map(|(&p, set)| if set.is_empty() { Some(p) } else { None })
+            .collect::<Vec<_>>();
+        for p in to_remove.iter() {
+            self.v1.remove(p);
+        }
+        // Add these bins.
+        let mut bins = bins.clone();
+        self.v1
+            .entry(pkg.package_id())
+            .and_modify(|set| set.append(&mut bins))
+            .or_insert(bins);
+    }
+
+    fn remove(&mut self, pkg_id: PackageId, bins: &BTreeSet<String>) {
+        let mut installed = match self.v1.entry(pkg_id) {
+            btree_map::Entry::Occupied(e) => e,
+            btree_map::Entry::Vacant(..) => panic!("v1 unexpected missing `{}`", pkg_id),
+        };
+
+        for bin in bins {
+            installed.get_mut().remove(bin);
+        }
+        if installed.get().is_empty() {
+            installed.remove();
+        }
+    }
+
+    fn save(&self, lock: &FileLock) -> CargoResult<()> {
+        let mut file = lock.file();
+        file.seek(SeekFrom::Start(0))?;
+        file.set_len(0)?;
+        let data = toml::to_string(self)?;
+        file.write_all(data.as_bytes())?;
+        Ok(())
     }
 }
 
+impl CrateListingV2 {
+    fn sync_v1(&mut self, v1: &CrateListingV1) -> CargoResult<()> {
```

Comment:
So in the long term we only need one upgrade path I think, right? We need one piece of logic which promotes a v1 manifest into a v2 manifest, and then we'd presumably zero out or delete the v1 manifest afterwards. 

Is the purpose of this function then to sort of keep these in sync during the transitionary phase leading up to that glorious future? It's a little unclear what the internals are doing here to me and why this is written the way that it is, but a comment on the function may also help with docs about intended usage patterns!

Reply:
Yea, it is intended that there would be a transition period if/when it is stabilized. At some point in the future, v1 can be removed. It may seem like a lot of complexity, but I wanted to make sure the transition goes smoothly. I've added some comments, but in general the idea is that v1 is the truth, and v2 always picks up changes from v1.

Reply:
Ok this sounds like a solid strategy to me, and the comment added here definitely clarifies it as well, thanks!